Sources/SentryCrash/Recording/Tools/SentryHook.c

```diff
@@ -0,0 +1,227 @@
+#include "SentryHook.h"
+#include "fishhook.h"
+#include <dispatch/dispatch.h>
+#include <execinfo.h>
+#include <mach/mach.h>
+#include <pthread.h>
+
+// NOTE on accessing thread-locals across threads:
+// We save the async stacktrace as a thread local when dispatching async calls,
+// but the various crash handlers need to access these thread-locals across threads
+// sometimes, which we do here:
+// While `pthread_t` is an opaque type, the offset of `thread specific data` (tsd)
+// is fixed due to backwards compatibility.
+// See:
+// https://github.com/apple/darwin-libpthread/blob/c60d249cc84dfd6097a7e71c68a36b47cbe076d1/src/types_internal.h#L409-L432
+
+#if __LP64__
+#    define TSD_OFFSET 224
+#else
+#    define TSD_OFFSET 176
+#endif
+
+static pthread_key_t async_caller_key = 0;
+
+sentrycrash_async_backtrace_t *
+sentrycrash_get_async_caller_for_thread(SentryCrashThread thread)
+{
+    const pthread_t pthread = pthread_from_mach_thread_np((thread_t)thread);
+    void **tsd_slots = (void *)((uint8_t *)pthread + TSD_OFFSET);
+
+    return (sentrycrash_async_backtrace_t *)__atomic_load_n(
+        &tsd_slots[async_caller_key], __ATOMIC_SEQ_CST);
+}
+
+void
+sentrycrash__async_backtrace_incref(sentrycrash_async_backtrace_t *bt)
+{
+    if (!bt) {
+        return;
+    }
+    __atomic_fetch_add(&bt->refcount, 1, __ATOMIC_SEQ_CST);
+}
+
+void
+sentrycrash__async_backtrace_decref(sentrycrash_async_backtrace_t *bt)
+{
+    if (!bt) {
+        return;
+    }
+    if (__atomic_fetch_add(&bt->refcount, -1, __ATOMIC_SEQ_CST) == 1) {
+        sentrycrash__async_backtrace_decref(bt->async_caller);
+        free(bt);
+    }
+}
+
+sentrycrash_async_backtrace_t *
+sentrycrash__async_backtrace_capture(void)
+{
+    sentrycrash_async_backtrace_t *bt = malloc(sizeof(sentrycrash_async_backtrace_t));
+    bt->refcount = 1;
+
+    bt->len = backtrace(bt->backtrace, MAX_BACKTRACE_FRAMES);
+
+    sentrycrash_async_backtrace_t *caller = pthread_getspecific(async_caller_key);
+    sentrycrash__async_backtrace_incref(caller);
+    bt->async_caller = caller;
+
+    return bt;
+}
+
+static void (*real_dispatch_async)(dispatch_queue_t queue, dispatch_block_t block);
+
+void
+sentrycrash__hook_dispatch_async(dispatch_queue_t queue, dispatch_block_t block)
+{
+    // create a backtrace, capturing the async callsite
+    sentrycrash_async_backtrace_t *bt = sentrycrash__async_backtrace_capture();
+
+    return real_dispatch_async(queue, ^{
```

Comment:
Is there a chance this never gets scheduled and `bt` leaks?

Same for `real_dispatch_after` and `real_dispatch_barrier_async` below.

Reply:
I think its possible in theory. While the pointer is moved into the block, it doesn’t know how to drop it. I don’t think its a problem in reality, or I would guess a lot more software would have similar leak issues.